symengine/series_generic.cpp

```diff
@@ -91,66 +96,160 @@ std::string UnivariateSeries::__str__() const
         }
         first = false;
         if (it.first == 0) {
-            o << mp_abs(it.second);
+            o << Expression(abs(it.second.get_basic()));
             continue;
         }
-        if (mp_abs(it.second) == 1)
-            o << var_->get_name();
+        if (Expression(abs(it.second.get_basic())) == 1)
+            o << var_;
         else
-            o << mp_abs(it.second) << "*" << var_->get_name();
+            o << Expression(abs(it.second.get_basic())) << "*" << var_;
         if (it.first > 1)
             o << "**" << it.first;
     }
     if (o.str() != "0")
-        o << " + O(" << var_->get_name() << "**" << prec_ << ")";
+        o << " + O(" << var_ << "**" << prec_ << ")";
     return o.str();
 }
 
-RCP<const UnivariateSeries> add_uni_series (const UnivariateSeries& a, const UnivariateSeries& b)
-{
-    map_uint_mpz dict;
-    SYMENGINE_ASSERT(a.var_->get_name() == b.var_->get_name())
+RCP<const UnivariatePolynomial> UnivariateSeries::convert_vector(const std::vector<integer_class> &v) {
+    std::vector<Expression> vtrunc;
+    for (const auto &it : v)
+        if (mp_get_si(it) < prec_) 
+            vtrunc.push_back(mp_get_si(it));
+    return UnivariatePolynomial::create(symbol(var_), vtrunc);
+}
+
+unsigned UnivariateSeries::ldegree(const UnivariateExprPolynomial &s) {
+    return s.get_univariate_poly()->get_dict().begin()->first;
+}
+
+UnivariateExprPolynomial UnivariateSeries::mul(const UnivariateExprPolynomial &s, const UnivariateExprPolynomial &r, unsigned prec) {
+    // No prec mul
+    return s * r;
+}
+
+UnivariateExprPolynomial UnivariateSeries::pow(const UnivariateExprPolynomial &s, int n, unsigned prec) {
+    // No prec pow
+    return pow_poly(s, n);
+}
+
+Expression UnivariateSeries::find_cf(const UnivariateExprPolynomial &s, const UnivariateExprPolynomial &var, unsigned deg) {
+    return (s.get_univariate_poly()->get_dict()).at(deg);
+}
+
+Expression UnivariateSeries::root(Expression &c, unsigned n) {
+    return pow_ex(c, 1/Expression(n));
+}
+
+UnivariateExprPolynomial UnivariateSeries::diff(const UnivariateExprPolynomial &s, const UnivariateExprPolynomial &var) {
+    return UnivariateExprPolynomial(Expression(s.get_univariate_poly()->diff(var.get_univariate_poly()->get_var())));
```

Comment:
This should get the value `s.get_univariate_poly()->diff(var.get_univariate_poly()->get_var())` and check if it is a `UnivariatePolynomial` and use the `UnivariatePolynomial` constructor. If not throw an error